pkg/kernel/stage.go

```diff
@@ -76,61 +87,81 @@ func (s *stage) run(k *kernel) {
 	}
 
 	s.running.Signal()
+
+	return s.waitUntilHealthy()
 }
 
-func (s *stage) stopWait(stageIndex int, logger log.Logger) {
-	s.cfn.Kill(ErrKernelStopping)
-	s.err = s.cfn.Wait()
+func (s *stage) healthcheck() HealthCheckResult {
+	var ok bool
+	var err error
+	var healthAware HealthCheckedModule
+	var result HealthCheckResult
+
+	for name, ms := range s.modules.modules {
+		if healthAware, ok = ms.module.(HealthCheckedModule); !ok {
+			continue
+		}
 
-	if s.err != nil && s.err != ErrKernelStopping {
-		logger.Error("error during the execution of stage %d: %w", stageIndex, s.err)
+		ok, err = healthAware.IsHealthy(s.ctx)
+
+		result = append(result, ModuleHealthCheckResult{
+			StageIndex: s.index,
+			Name:       name,
+			Healthy:    ok,
+			Err:        err,
+		})
 	}
 
-	s.terminated.Signal()
+	return result
 }
 
-func (s *stage) len() int {
-	return s.modules.len()
-}
+func (s *stage) waitUntilHealthy() error {
+	var result HealthCheckResult
+
+	timeoutTicker := clock.NewRealTimer(s.healthCheckSettings.Timeout)
```

Comment:
Should you stop these tickets when leaving the function (aka add a defer calling stop on both)? Otherwise we will leak them

Reply:
added the defer statements :+1: